src/compiler/checker.ts

```diff
@@ -22149,6 +22172,25 @@ namespace ts {
                 && (!roundTripOnly || s === pseudoBigIntToString({ negative, base10Value: parsePseudoBigInt(scanner.getTokenValue()) }));
         }
 
+        function isMemberOfStringMapping(source: Type, target: Type): boolean {
+            if (target.flags & (TypeFlags.String | TypeFlags.AnyOrUnknown)) {
+                return true;
+            }
+            if (target.flags & TypeFlags.TemplateLiteral) {
+                return isTypeAssignableTo(source, target);
+            }
+            if (target.flags & TypeFlags.StringMapping) {
+                const mappingStack = [];
+                while (target.flags & TypeFlags.StringMapping) {
+                    mappingStack.unshift(target.symbol);
+                    target = (target as StringMappingType).type;
+                }
+                const mappedSource = reduceLeft(mappingStack, (memo, value) => getStringMappingType(value, memo), source);

```

Comment:
You could just use `push` and `reduceRight`, no? I guess we might not define our own `reduceRight`.

Also, would be nice to have a comment. Consider:

```suggestion
                // We need to see whether applying the same mappings of the target
                // onto the source would produce an identical type *and* that
                // it's compatible with the inner-most non-string-mapped type.
                //
                // The intuition here is that if same mappings don't affect the source at all,
                // and the source is compatible with the unmapped target, then they must
                // still reside in the same domain.
                const mappingStack = [];
                while (target.flags & TypeFlags.StringMapping) {
                    mappingStack.push(target.symbol);
                    target = (target as StringMappingType).type;
                }
                const mappedSource = mappingStack.reduceRight((memo, value) => getStringMappingType(value, memo), source);
```

Reply:
I agree, a comment like that would make it clearer what's going on.

Reply:
> You could just use push and reduceRight, no? I guess we might not define our own reduceRight.

Yep, that's the reason. We only have a `reduceLeft` and not a `reduceRight` utility function, and I usually prefer to use those if they're available (otherwise we should remove them right?). Given this array will likely never be more than a handful of elements long, my preference would be to keep the `core` helper in use, rather than use the (es6) array method. Less work for that one guy patching the TS compiler to run on a blackberry phone browser.